File: lang/java/reef-examples/src/main/java/org/apache/reef/examples/hello/HelloREEF.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.reef.examples.hello;

import org.apache.reef.client.DriverConfiguration;
import org.apache.reef.client.DriverLauncher;
import org.apache.reef.client.LauncherStatus;
import org.apache.reef.runtime.local.client.LocalRuntimeConfiguration;
import org.apache.reef.tang.Configuration;
import org.apache.reef.tang.exceptions.InjectionException;
import org.apache.reef.util.EnvironmentUtils;
import org.apache.reef.util.ThreadLogger;

import java.util.logging.Level;
import java.util.logging.Logger;

/**
 * The Client for Hello REEF example.
 */
public final class HelloREEF {

  private static final Logger LOG = Logger.getLogger(HelloREEF.class.getName());

  /** Number of milliseconds to wait for the job to complete. */
  private static final int JOB_TIMEOUT = 10000; // 10 sec.


  /** Configuration of the runtime. */
  private static final Configuration RUNTIME_CONFIG =
      LocalRuntimeConfiguration.CONF
          .set(LocalRuntimeConfiguration.MAX_NUMBER_OF_EVALUATORS, 2)
          .build();

  /** Configuration of the HelloREEF driver. */
  private static final Configuration DRIVER_CONFIG =
      DriverConfiguration.CONF
          .set(DriverConfiguration.DRIVER_IDENTIFIER, "HelloREEF")
          .set(DriverConfiguration.GLOBAL_LIBRARIES, EnvironmentUtils.getClassLocation(HelloDriver.class))
          .set(DriverConfiguration.ON_DRIVER_STARTED, HelloDriver.StartHandler.class)
          .set(DriverConfiguration.ON_EVALUATOR_ALLOCATED, HelloDriver.EvaluatorAllocatedHandler.class)
          .build();

  /**
   * Start Hello REEF job with local runtime.
   * @param args command line parameters.
   * @throws InjectionException configuration error.
   */
  public static void main(final String[] args) throws InjectionException {

    final LauncherStatus status = DriverLauncher.getLauncher(RUNTIME_CONFIG).run(DRIVER_CONFIG, JOB_TIMEOUT);

    LOG.log(Level.INFO, "REEF job completed: {0}", status);

    ThreadLogger.logThreads(LOG, Level.FINE, "Threads running at the end of HelloREEF:");
  }

  /** Empty private constructor to prohibit instantiation of utility class. */
  private HelloREEF() { }
}

File: lang/cs/Org.Apache.REEF.Bridge.Core.HelloREEF/HelloREEF.cs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

using System;
using System.Globalization;
using Org.Apache.REEF.Bridge.Core.Common.Client.Config;
using Org.Apache.REEF.Bridge.Core.Common.Client.Config.Runtime;
using Org.Apache.REEF.Bridge.Core.Grpc.Client;
using Org.Apache.REEF.Tang.Annotations;
using Org.Apache.REEF.Tang.Implementations.Tang;
using Org.Apache.REEF.Tang.Interface;
using Org.Apache.REEF.Tang.Util;
using Org.Apache.REEF.Utilities.Logging;

namespace Org.Apache.REEF.Bridge.Core.HelloREEF
{
    /// <summary>
    /// A Tool that submits HelloREEFDriver for execution.
    /// </summary>
    public sealed class HelloREEF
    {
        private static readonly Logger Log = Logger.GetLogger(typeof(HelloREEF));

        private const string Local = "local";
        private const string YARN = "yarn";
        private const string HDInsight = "hdi";
        private const string AzureBatch = "azurebatch";

        [Inject]
        private HelloREEF()
        {
        }

        /// <summary>
        /// Runs HelloREEF using the IREEFClient passed into the constructor.
        /// </summary>
        private void Run(string runtime)
        {
            // The driver configuration contains all the needed bindings.
            var helloDriverConfiguration = DriverApplicationConfiguration.ConfigurationModule
                .Set(DriverApplicationConfiguration.OnDriverStarted, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.OnDriverStopped, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.OnEvaluatorAllocated, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.OnEvaluatorCompleted, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.OnEvaluatorFailed, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.OnContextActive, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.OnTaskRunning, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.OnTaskCompleted, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.OnTaskFailed, GenericType<HelloDriver>.Class)
                .Set(DriverApplicationConfiguration.CustomTraceLevel, Level.Verbose.ToString())
                .Build();

            var applicationId = GetApplicationId();

            // The JobSubmission contains the Driver configuration as well as the files needed on the Driver.
            var driverRuntimeConfiguration = DriverRuntimeConfiguration.ConfigurationModule
                .Set(DriverRuntimeConfiguration.OsType, GenericType<OsLinux>.Class)
                .Set(DriverRuntimeConfiguration.JobId, applicationId);
            driverRuntimeConfiguration = DriverRuntimeConfiguration.AddGlobalAssemblyForType(driverRuntimeConfiguration, typeof(HelloDriver));

            var runtimeConfiguration = GetRuntimeConfiguration(runtime);

            using (var launcher = ClientLauncherFactory.GetLauncher(runtimeConfiguration, driverRuntimeConfiguration.Build()))
            {
                var launcherStatus = launcher.SubmitAsync(helloDriverConfiguration).Result;
                Log.Log(Level.Info, "Final Launch Status {0}", launcherStatus);
                if (launcherStatus.Error.IsPresent())
                {
                    throw launcherStatus.Error.Value;
                }
            }
        }

        private string GetApplicationId()
        {
            return "HelloWorldJob-" + DateTime.Now.ToString("ddd-MMM-d-HH-mm-ss-yyyy", CultureInfo.CreateSpecificCulture("en-US"));
        }

        /// <summary>
        /// </summary>
        /// <param name="name"></param>
        /// <returns></returns>
        private static IConfiguration GetRuntimeConfiguration(string name)
        {
            switch (name)
            {
                case Local:
                    return LocalRuntimeConfiguration.ConfigurationModule
                        .Set(LocalRuntimeConfiguration.NumberOfEvaluators, "2")
                        .Build();
                case YARN:
                    return YarnRuntimeConfiguration.ConfigurationModule.Build();
                case HDInsight:
                    // To run against HDInsight please replace placeholders below, with actual values for
                    // connection string, container name (available at Azure portal) and HDInsight
                    // credentials (username and password)
                    const string connectionString = "ConnString";
                    const string continerName = "foo";
                    return HdInsightRuntimeConfiguration.ConfigurationModule
                        .Set(HdInsightRuntimeConfiguration.HdInsightPassword, @"pwd")
                        .Set(HdInsightRuntimeConfiguration.HdInsightUserName, @"foo")
                        .Set(HdInsightRuntimeConfiguration.HdInsightUrl, @"https://foo.azurehdinsight.net/")
                        .Build();
                case AzureBatch:
                    return AzureBatchRuntimeConfiguration.ConfigurationModule
                        .Set(AzureBatchRuntimeConfiguration.AzureBatchAccountKey, "###")
                        .Set(AzureBatchRuntimeConfiguration.AzureBatchAccountName, "###")
                        .Set(AzureBatchRuntimeConfiguration.AzureBatchAccountUri, "###")
                        .Set(AzureBatchRuntimeConfiguration.AzureBatchPoolId, "###")
                        .Set(AzureBatchRuntimeConfiguration.AzureStorageAccountKey, "###")
                        .Set(AzureBatchRuntimeConfiguration.AzureStorageAccountName, "###")
                        .Set(AzureBatchRuntimeConfiguration.AzureStorageContainerName, "###")
                        .Build();
                default:
                    throw new Exception("Unknown runtime: " + name);
            }
        }

        public static void MainSimple(string[] args)
        {
            var runtime = args.Length > 0 ? args[0] : Local;

            // Execute the HelloREEF, with these parameters injected
            TangFactory.GetTang()
                .NewInjector()
                .GetInstance<HelloREEF>()
                .Run(runtime);
        }
    }
}
